policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/service/PolicyServiceHierarchyProxy.java

```diff
@@ -42,101 +41,174 @@
+     * data type to extract and merge the policies at each stage of the hierarchy.
+     *
+     * @param resource       A {@link Resource} to get the applicable rules for.
+     * @param rulesExtractor The rule type to extract from each policy
+     * @param <R>            the type of resource (may be a supertype)
+     * @return An optional {@link Rules} object, which contains the list of rules found
+     * that need to be applied to the resource.
+     */
+    private <R extends Resource> Optional<Rules<LeafResource>> getResourceRules(final R resource, final Function<Resource, Optional<Rules<LeafResource>>> rulesExtractor) {
+        LOGGER.debug("Getting the applicable rules: {}", resource);
+        Optional<Rules<LeafResource>> inheritedRules;
+        if (resource instanceof ChildResource) {
+            // We will also need the policy applied to the parent resource
+            LOGGER.debug("resource {} an instance of ChildResource", resource);
+            inheritedRules = getResourceRules(((ChildResource) resource).getParent(), rulesExtractor);
+            LOGGER.debug("Inherited rules {} for resource {}", inheritedRules, resource);
+        } else {
+            // We are at top of hierarchy
+            LOGGER.debug("resource {} NOT an instance of ChildResource (top of hierarchy)", resource);
+            inheritedRules = Optional.empty();
         }
-        mergedRules.message(String.join(",", messages));
-        LOGGER.debug("Merged messages: {} + {} -> {}", inheritedRules.getMessage(), newRules.getMessage(), mergedRules.getMessage());
 
-        // Merge rules
-        mergedRules.addRules(inheritedRules.getRules());
-        mergedRules.addRules(newRules.getRules());
-        LOGGER.debug("Merged rules: {} + {} -> {}", inheritedRules.getRules(), newRules.getRules(), mergedRules.getRules());
+        Optional<Rules<LeafResource>> newRules = rulesExtractor.apply(resource);
 
-        return mergedRules;
+        // If both present, merge both
+        // If either present, return present
+        // If none present, return Optional.empty()
+        return inheritedRules.map(iRules -> newRules.map(nRules -> mergeRules(iRules, nRules)).or(() -> inheritedRules)).orElse(newRules);
+    }
+
+    /**
+     * GetRecordRules is used by the service to get any record rules that could be applied against the resource that the user has requested
+     * If no rules are found then a {@link NoSuchPolicyException} will be thrown
+     *
+     * @param resource a {@link LeafResource} to get rules for
+     * @return the record rules that apply to the LeafResource
+     */
+    public Rules<Serializable> getRecordRules(final LeafResource resource) {
+        var optionalRules = getRecordRules(resource, service::getRecordRules);
+
+        return optionalRules
+                .filter(rules -> !rules.getRules().isEmpty())
+                .orElseThrow(() -> new NoSuchPolicyException("No Policy Found"));
```

Comment:
Would be much more useful to have more info
No *record* rules found for this leafResource perhaps?

Reply:
updated